File: AdaBoost/myAdaBoostRT.R
AdaBoost.RT<-function ( x, y, x.test=NULL, method='mlp', phi=0.5, maxIter=100, ...){
	# AdaBoost.RT: a Boosting Algorithm for Regression Problems
  #	D. P. Solomatine, D. L. Shrestha
	
	m<-nrow(x)
	n<-ncol(x)
	y<-matrix(y,nrow=m)
	
	T=maxIter # max iter
	#initialize weight distr and error
	epsilon<-beta<-D<-rep(NA,T)
	fit<-e<-w<-array(NA,dim=c(T,m))
	w[1,]=rep(1/m, m)
	
	# iterate
	t=1
	while (t<T){
		idx<-sample(m,replace=TRUE,prob=w[t,])
		tth_x=x[idx,]
		tth_y=y[idx,]
		
		# call weak learner, CART
		model[[t]]=eval(parse(text=method))(tth_x,tth_y,...)
		# getback to hypothesis
		fit[t,]<-y_fit<-predict(model[[t]],x)
		# adjusted error for each instance
		ARE<-abs((y-y_fit)/y)
		is_ARE_gt_phi=ARE>phi
		is_ARE_le_phi=ARE<=phi
		
		# calculate error of hypothesis
		epsilon[t]=sum(w[t,is_ARE_gt_phi])

		# calculate beta
		beta[t]<-epsilon[t]^2
		# update weight vector
		w[t+1,is_ARE_le_phi]=w[t,is_ARE_le_phi]*beta[t]
		w[t+1,is_ARE_gt_phi]=w[t,is_ARE_gt_phi]
		# normalize
		w[t+1,]=w[t+1,]/sum(w[t+1,])
		t=t+1
	}
	max_t=t-1
	
	# final model
	ww=log(1/beta[1:max_t])
	
	aggregated.fit<-ww%*%fit[1:max_t,]/sum(ww)
	
	# predict
	if(!is.null(x.test)){ # have test dataset
		m.test<-nrow(x.test)
		predict<-array(NA,dim=c(max_t,m.test))
		for (t in 1:max_t){
			predict[t,]<-predict(model[[t]],x.test)
		}
		aggregated.predict<-ww%*%predict/sum(ww)
	}else{
		aggregated.predict=NULL
	}
	return(list(aggregated.fit=aggregated.fit,aggregated.predict=aggregated.predict,
							base.fit=fit,base.predict=predict,beta=beta[1:max_t],weight=ww,max.t=max_t))
}

File: matlab-src/myAdaBoostRT.m
function [ z_pred, beta ] = myAdaBoostRT( trn_fun, tst_fun, x, y, z, maxIter, phi, pcoef )
% Experiments with a New Boosting Algorithm
% Yoav Freund Robert E. Schapire
% pcoef=2;

% example 1
% load moore
% y = moore(:,6);              % Response
% x = moore(:,1:5);            % Original predictors
% z = x+10*randn(size(x));       % Correlated predictors
% phi=0.5;
% maxIter=50;
% pcoef=2;
% [ z_pred, beta ] = myAdaBoostRT( x, y, z, maxIter, phi, pcoef );

[m,n]=size(x);
%initialize weight distr and error
D_new=repmat(1/m, m, 1);

T=maxIter; % max iter
% phi=0.5; % error threshold

e=zeros(1,T);
beta=ones(1,T);
f=cell(1,T);
% iterate
t=1;
while t<=T
    D=D_new;
    % bootstrap with dist D
    [tmp,bootsam]=bootstrp(1, @mean, x, 'Weights',D);
    tth_x=x(bootsam,:);
    tth_y=y(bootsam,:);
    % call weak learner, CART
    f{t}=trn_fun(tth_x,tth_y);
    % getback to hypothesis
    y_fit=tst_fun(f{t},x);
    ARE=abs((y_fit-y)./y);
    is_are_lt_phi=ARE>phi;
    
    % calculate error rate
    e(t)=D'*is_are_lt_phi;
    if(e(t)==0) % perfect bypass this one
        continue;
    end
    % question!!!
    % if all correct (perfect predict), then e=0 and beta=0
    %     th_e=e(t);
    % weight updating param
    beta(t)=e(t).^pcoef;
    % update distr
    lt_phi=beta(t)*(ARE<=phi);
    ge_phi=is_are_lt_phi;
    D_new=D.*(lt_phi+ge_phi);
    % normalize
    D_new=D_new./sum(D_new);
    t=t+1;
end
max_t=t-1;
% final model: cart, beta(t)
ww=log(1./beta(1:max_t));

% test
pred=zeros(size(z,1), max_t);
for t=1:max_t
    pred(:,t)=tst_fun(f{t},z);
end

norminator=ww*pred';
denorminator=sum(ww);

z_pred=norminator./denorminator;
z_pred=z_pred';
end



